src/spu/src/services/public/stream_fetch.rs

```diff
@@ -169,96 +169,86 @@ impl StreamFetchHandler {
-                        }
+                    // consume hasn't read all offsets, need to send back gaps
+                    debug!(
+                        changed_consumer_offset,
+                        last_read_offset,
+                        "need send back"
+                    );
+                    let (offset, wait) = self.send_back_records(changed_consumer_offset, module.as_ref()).await?;
+                    last_read_offset = offset;
+                    if wait {
+                        consumer_offset = None;
+                        debug!(
+                            last_read_offset,
+                            "wait for consumer"
+                        );
+                    } else {
+                        consumer_offset = Some(last_read_offset);   // no need wait for consumer, skip it
                     }
                 },
 
                 // received new offset from leader
                 leader_offset_update = leader_offset_receiver.listen() => {
-
                     debug!(leader_offset_update);
 
-                    if let Some(last_consumer_offset) = consumer_offset {
-                        // we know what consumer offset is
-                        if leader_offset_update > last_consumer_offset {
-                            debug!(leader_offset_update,
-                                consumer_offset = last_consumer_offset,
-                                "reading offset event");
-                            let (offset,wait) = self.send_back_records(last_consumer_offset,module.as_ref()).await?;
-                            last_read_offset = offset;
-                            if wait {
-                                consumer_offset = None;
-                                debug!(
-                                    last_read_offset,
-                                    "wait for consumer"
-                                );
-                            } else {
-                                consumer_offset = Some(last_read_offset);   // no need wait for consumer, skip it
-                            }
-
-
-                        } else {
-                            debug!(ignored_update_offset = leader_offset_update);
+                    let last_consumer_offset = match consumer_offset {
+                        Some(last_consumer_offset) => last_consumer_offset,
+                        None => {
+                            // we don't know consumer offset, so we delay
+                            debug!(delay_consumer_offset = leader_offset_update);
                             last_read_offset = leader_offset_update;
-                        }
-                    } else {
-                        // we don't know consumer offset, so we delay
-                        debug!(delay_consumer_offset = leader_offset_update);
-                        last_read_offset = leader_offset_update;
+                            continue;
+                        },
+                    };
```

Comment:
Here, reducing right-drift by using a match statement to unwrap `last_consumer_offset`. If the value is `None`, just `continue`.